packages/cli-server-api/src/openURLMiddleware.ts

```diff
@@ -31,20 +31,37 @@ async function openURLMiddleware(
 
     const {url} = req.body as {url: string};
 
+    if (typeof url !== 'string') {
+      res.writeHead(400);
+      res.end('URL must be a string');
+      return;
+    }
+
+    let parsedUrl: URL;
     try {
-      const parsedUrl = new URL(url);
-      if (parsedUrl.protocol !== 'http:' && parsedUrl.protocol !== 'https:') {
-        res.writeHead(400);
-        res.end('Invalid URL protocol');
-        return;
-      }
+      parsedUrl = new URL(url);
     } catch (error) {
       res.writeHead(400);
       res.end('Invalid URL format');
       return;
     }
 
-    await open(url);
+    if (parsedUrl.protocol !== 'http:' && parsedUrl.protocol !== 'https:') {
+      res.writeHead(400);
+      res.end('Invalid URL protocol');
+      return;
+    }
+
+    // Reconstruct URL with proper encoding to prevent command injection
+    // The URL constructor doesn't automatically encode special characters like | in query strings,
```

Comment:
To be specific, it encodes special characters, but only sets of them in each URL part [^0]. For example, `|` is encoded in [`userinfo`](https://url.spec.whatwg.org/#userinfo-percent-encode-set):
```js
new URL('https://user|:pass@example.com')`
// https://user%7C:pass@example.com/
```

Current implementation double-encodes several characters for that reason; for example, whitespaces:
```js
const parsedUrl = new URL('https://example.com/?#some hash')
// https://example.com/?#some%20hash
const sanitizedUrl = new URL(parsedUrl.origin);
// ...
console.log(sanitizedUrl.href)
// https://example.com/#some%2520hash
```

A simpler approach could be:
```js
const sanitizedUrl = encodeURI(url);
```

[^0]: https://url.spec.whatwg.org/#percent-encoded-bytes